Add zgenhostid utility script and package with utilities

ofaaland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Description
<!--- Describe your changes in detail -->
This commit adds a script modeled after genhostid(1).

Zgenhostid checks for an /etc/hostid file; if it does not exist, it
creates one and stores a value.  If the user has provided a hostid as an
argument, that value is used.  Otherwise, a random hostid is generated
and stored.

This differs from the CENTOS 6/7 versions of genhostid, which overwrite
the /etc/hostid file even though their manpages state otherwise.

Add a man page. The one for genhostid is in (1), but I put zgenhostid in
(8) because I believe it's more appropriate.

Modified mmp tests to use zgenhostid to set the hostid instead of using
the spl_hostid module parameter.  zgenhostid will not replace an
existing /etc/hostid file, so new mmp_clear_hostid calls are required.

Fixes #6358

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Turning the multihost property on requires that a hostid be set to allow
ZFS to determine when a foreign system is attemping to import a pool.
The error message instructing the user to set a hostid refers to
genhostid(1).

Genhostid(1) is not available on SUSE Linux.  

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

ofaaland commented:
@adilger, any feedback would be welcome.  Thanks

adilger commented on cmd/zpool/zpool_main.c:
```diff
@@ -2177,7 +2177,8 @@ show_import(nvlist_t *config)
 		case ZPOOL_STATUS_HOSTID_REQUIRED:
 			(void) printf(gettext(" action: Set a unique system "
-			    "hostid with the genhostid(1) command.\n"));
+			    "hostid with the genhostid(1) or zgenhostid(8) "
```
Since you are providing the zgenhostid(8) command, you might consider to just reference that exclusively, and install it for all distros?

ofaaland replied:
agreed

adilger commented on man/man8/zgenhostid.8:
```diff
@@ -0,0 +1,28 @@
+Specifies the value to be placed in \fB/etc/hostid\fR.  It must be a
+number with a value between 1 and 2^32-1.  This value \fBmust\fR be
+unique among your systems.  It must be expressed in hexadecimal and be
+exactly 8 digits long.
```
I did confirm that the existing `hostid` command will print the leading zero, so it will indeed always be 8 digits long

ofaaland replied:
Yes, I checked that too.

adilger commented on man/man8/zpool.8:
```diff
@@ -732,7 +732,7 @@ in the
 See
 .Xr genhostid 1
-and
+.Xr zgenhostid 1
```
`.Xr zgenhostid 8` ?

ofaaland replied:
fixed, thanks

adilger commented:
The upstream issue tracker for `genhostid` overwriting an existing `/etc/hostid` file is being tracked at https://github.com/fedora-sysv/initscripts/issues/105

ofaaland commented on cmd/zgenhostid/zgenhostid:
```diff
@@ -0,0 +1,60 @@
+# which do not have that utility.
+#
+# Usage:
+#    genhostid
```
fixed

ofaaland commented on cmd/zgenhostid/zgenhostid:
```diff
@@ -0,0 +1,60 @@
+#    genhostid
+#    genhostid <value>
+#
+# If /etc/hostid already exists, the script exits immediately and changes
```
OK.  I'll go with the nonzero size.

ofaaland commented on cmd/zgenhostid/zgenhostid:
```diff
@@ -0,0 +1,60 @@
+	echo "be generated.  The value must be unique among your systems." >&2
+}
+
+if [ -e $hostid_file ]; then
```
agreed

ofaaland commented on cmd/zgenhostid/zgenhostid:
```diff
@@ -0,0 +1,60 @@
+	# $RANDOM goes from 0..32k-1
+	number1=$(( ( (RANDOM%4)*32768*32768 ) ))
+	number2=$(( ( RANDOM*32768) ))
+	number3=$(( ( RANDOM) ))
```
That's a good point that the libc-generated one may already be in use.  

My concern is that someone will either have a poorly chosen name-to-address mapping, so that two machines end up with the same libc-generated hostid; or they have something like corosync in place that changes the IP address as part of a failover mechanism.

Do you think those are not meaningful risks?

And, it's already possible to get the same effect with:
`
zgenhostid $(hostid)
`
Although that should probably be in the manpage.

ofaaland commented:
> The upstream issue tracker for genhostid overwriting an existing /etc/hostid file is being tracked at fedora-sysv/initscripts#105
Thanks for doing that.

ofaaland commented:
I pushed an update that addressed the review comments.  It still generates a random hostid by default rather than using the one provided by libc, but I'm still thinking over that proposal.

ofaaland commented:
Update pushed after Andreas' approval is just a squash and rebase on master.  No code change to the patch.

ofaaland commented:
Supersed by #6397
